File: algorithms/graph.cpp
#include <map>
#include <queue>
#include <string>
#include <vector>

typedef std::pair<std::string, std::vector<std::string>> dict;

void push_all(std::vector<std::string> ppl, std::queue<std::string> *q){
    for (const auto& p: ppl){
        q->push(p);
    }
}

bool search_exists(std::vector<std::string> s, std::string name) {
    for(const auto& c: s){
        if (c == name) return true;
    }
    return false;
}

void check(std::string name){
    printf("check for %s...", name.c_str());
}

int main() {
    std::string seller = "Ken";
    std::map<std::string, std::vector<std::string>> people{
        {"you", {"Alice", "Tom"}},
        {"Alice", {"Magie"}},
        {"Tom", {"Daniel", "Clara"}},
        {"Clara", {"Alice", "Ken"}},
        {"Magie", {"you"}},
        {"Daniel", {}},
        {"Ken", {}}
    };

    std::queue<std::string> search_engine;
    push_all(people.at("you"), &search_engine);
    std::vector<std::string> search_ppl;

    while (!search_engine.empty()){
        std::string p = search_engine.front();
        search_engine.pop();
        if (search_exists(search_ppl, p)) continue;
        check(p);
        if (p == seller){
            printf("%s is a seller\n", p.c_str());
            return 0;
        }
        push_all(people.at(p), &search_engine);
        search_ppl.push_back(p);
    }
    return 0;
}

File: algorithms/graph.py
from collections import deque

def BFS():
    seller = "Clara"

    graph = {}
    graph["you"] = ["Alice", "Tom"]
    graph["Alice"] = ["Magie", "you"]
    graph["Tom"] = ["Alice", "Daniel", "Clara"]
    graph["Magie"] = ["Tom", "Alice"]
    graph["Daniel"] = []
    graph["Clara"] = []

    search_engine = deque()
    search_engine += graph["you"]
    searched_people = []

    while search_engine:
        person = search_engine.popleft()
        print("search for ", person, "...")
        if person not in searched_people:
            if person == seller:
                print (person + " is a seller")
                return True
            print("add ", person,"'s friends to a queue")
            search_engine += graph[person]
            searched_people.append(person)
        else:
            print(person, "was searched already")

    return False

BFS()


